Feat/#15 user login 기능 작성

heyoni opened the pull request:
## Definition
- #15 
- #24

heyoni committed:
feat: login 기능 추가

heyoni committed:
fix: JWT 발급시 불필요한 부분 수정

- subject(인증 외 다른용도로 사용시 작성) 삭제
- 함수 파라미터로 self 작성이 안되어있어서 추가함

f-lab-zion commented on schema/login.py:
```diff
@@ -0,0 +1,9 @@
+    access_token: str
+    token_type: str
+
+    class Config:
+        orm_mode = True
```
Login 클래스는 Response Model을 정의하기 위해 사용한 것으로 보입니다 `orm_mode=True` 로 설정한 이유가 있을까요?

heyoni replied:
orm_mode 관련 질문을 하고 지운다는걸 넣어버렸습니다! 수정하겠습니다!

f-lab-zion commented on services/LoginService.py:
```diff
@@ -0,0 +1,56 @@
+        encoded_jwt = jwt.encode(to_encode, settings.SECRET_KEY, algorithm=ALGORITHM)
+        return encoded_jwt
+
+    def login(self, form_data: OAuth2PasswordRequestForm = Depends()):
```
return type을 명시하면 좋겠습니다
또한 dict를 리턴하고 있는데 `Login` 이 되어야 하지 않을까요?

heyoni replied:
넵 수정했습니다~!

f-lab-zion commented on services/LoginService.py:
```diff
@@ -0,0 +1,56 @@
+    def create_access_token(self, expires_delta: timedelta = None) -> str:
+        if expires_delta:
+            expire = datetime.now() + expires_delta
+        else:
+            expire = datetime.now() + timedelta(
+                minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES
+            )
```
이 경우는 `create_access_token` 의 parameter 인 `expires_delta` 의 default 값을 `settings.ACCESS_TOKEN_EXPIRE_MINUTES` 로 정의하면 if-else 구문을 뺄 수 있습니다

heyoni replied:
감사합니다!

f-lab-zion commented on services/LoginService.py:
```diff
@@ -0,0 +1,56 @@
+from fastapi import Depends, HTTPException
+from jose import jwt
+
+ALGORITHM = "HS256"
+pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")
```
전역변수로 사용될 필요가 없어보입니다
LoginService 클래스 안으로 넣어도 되지 않을까요?

f-lab-zion commented on tests/test_login.py:
```diff
@@ -0,0 +1,73 @@
+
+    assert response.value.status_code == 401
+    assert response.value.detail == "아이디나 비밀번호가 틀렸습니다."
+
```
`username`, `password` 중 하나 또는 둘다 입력되지 않은 케이스에 대한 테스트 케이스가 추가되면 어떨까요?

f-lab-zion reviewed (changes requested):
👍 수고하셨습니다. 몇가지 코멘트만 확인해주세요 
그리고 login 으로 얻은 access_key를 사용하는 코드도 차후 다른 PR로 올려주세요

heyoni committed:
fix: 코드 리뷰 반영

heyoni committed:
feat: 추가 리뷰 반영(test 코드 수정)

- FakeRepository의 ID와 nickname값을 딕셔너리로 변경하여 딕셔너리의 get 메서드를 이용해서 원하는 값을 받아오게 함
- delete시 필요 없는 파라미터 변경
- conftest를 적용하여 반복적이거나 기본적인 구문들 정리

heyoni committed:
feat: Auth 추가, token paylode값 추가

- token이 있고 변경하려는 user와 요청하는 user가 같을 경우에만 Update, Delete가 가능하도록 변경함
- 해당 기능 test 추가
- token에 iss, sub, 추가

f-lab-zion commented on services/UserServices.py:
```diff
@@ -67,8 +74,17 @@ def update(self, id: int, user_body: UserUpdate) -> User:
         return self.userRepository.delete(id)
+
+    def get_current_user(self, token: str = Depends(oauth2_scheme)):
+        decoded_token = jwt.decode(token, settings.SECRET_KEY, algorithms=["HS256"])
+        req_user = self.get(int(decoded_token["sub"]))
```
잘못된 토큰이 오거나 토큰의 내용 중 `sub`키가 없는 경우를 대비하여 
Exception 처리를 해서 `raise HTTPException(...)` 해두면 어떨까요?

f-lab-zion commented on api/v1/endpoints/user.py:
```diff
@@ -1,12 +1,20 @@
 )
 
 
+def get_current_user(
+    userService: UserService = Depends(), token=Depends(oauth2_scheme)
+):
+    return userService.get_current_user(token)
```
지금은 user API만 있어서 여기에 둬도 괜찮습니다만 나중에 다른 API 에서도 이 함수를 Depends 하게 될텐데 함수 위치를 조정하는건 어떨까 싶습니다

heyoni replied:
상위 계층인 endpoint로 옮기는게 맞을까요??

f-lab-zion replied:
어떤 방식으로 정리할지는 정답으로 정해진건 없습니다 
만약 저라면 `/api/deps.py` 를 만들어서 `database.py` 에 있는 `get_db` 와 `api/v1/endpoints/user.py` 에 있는 `get_current_user` 를 함께 위치해두고 Depends()로 불러올 함수들을 한 군데 모아두는 것을 선호할 것 같습니다.
꼭 제 의견을 따를 필요는 없고 @heyoni 님이 생각하기에 좋은 방식으로 정리해두셔도 무방합니다

f-lab-zion commented on api/v1/endpoints/user.py:
```diff
@@ -18,11 +26,20 @@ def create_user(user_body: UserCreate, userService: UserService = Depends()):
+    id: int,
+    user_body: UserUpdate,
+    userService: UserService = Depends(),
+    req_user: str = Depends(get_current_user),
```
변수 Naming은 그 의미를 명확히 담고 있는게 좋다고 생각합니다 
`req_user` 보다는 `current_user` 가 좀더 명확하지 않을까요? (단순 제안입니다)

f-lab-zion reviewed (approved):
👍 좋아보입니다 
단순 코멘트만 몇개 달았으니 확인해보시고 머지하셔도 좋습니다 
수고하셨습니다

heyoni committed:
fix: 코드 리뷰 반영2

- 잘못된 토큰(sub가 없음)이 왔을 경우 예외 처리 추가
- Depends로 이용되는 함수들을 deps로 이동
- 변수 네이밍을 더 명확하게 변경
